README.md
<p align="center">
<img src="./imgs/logo.png"/>
</p>

<p>
  <a href="https://www.npmjs.com/package/socket-conveyor-client" target="_blank">
    <img alt="Version" src="https://img.shields.io/npm/v/socket-conveyor-client.svg">
  </a>
</p>


> A client for the php package kanata-php/socket-conveyor

### 🏠 [Homepage](https://github.com/kanata-php/socket-conveyor-client#readme)

## Prerequisites

- npm >=5.5.0
- node >=9.3.0

## Install

```sh
npm install socket-conveyor-client
```



## Description



This is a client for the [Socket Conveyor](https://github.com/kanata-php/socket-conveyor) PHP package..



## Usage



To use this package, once it is installed, the following example shows how it works.

```html
<div>
    <div><button onclick="connection.send('base', 'base-action')">Base Action</button></div>
    <div><button onclick="connection.send('first', 'example-first-action')">First Action</button></div>
    <div><button onclick="connection.send('second', 'example-second-action')">Second Action</button></div>
    <div id="output"></div>
</div>
<script type="module" type="text/javascript">

    // if this connection happens on a bundle, use it like this: import Conveyor from 'socket-conveyor-client';
    import Conveyor from './node_modules/socket-conveyor-client/index.js';

    var connection = new Conveyor({
        channel: 'actions-channel',
        listen: ['example-first-action'],
        onMessage: (e) => {
            document.getElementById('output').innerHTML = e;
        },
        onReady: () => {
            connection.send('Welcome!');
        },
    });

</script>
```

With the previous example your html client will try to connect to `ws://127.0.0.1:8000`. This client will connect to the channel `actions-channel`, and will listen to `example-first-action` actions. That said, any message sent to that **channel** and **action**, will be received at the `OnMessage` callback.

## Options

```
protocol: 'ws',
uri: '127.0.0.1',
port: 8000,

// Url query for connection.
query: '',

channel: null,
listen: null,
onOpen: this.onOpen.bind(this),
onReady: () => {},

// Message handler for only the data portion.
onMessage: () => {},

// Message handler for the whole incoming object.
onRawMessage: () => {},

// Callback triggered when closing connection. If overwritten the reconnection feature can be affected. 
onClose: () => this.onClose.bind(this),

// Callback for securely do something when connection is closed, without affecting default closing connection behavior.
onCloseCallback: () => {},

onError: () => {},

// Settings for reconnection feature.
reconnect: false,
reconnectDelay: 5000,

// Interval for connection verification, so reconnection action can take place if reconnection active.
healthCheckInterval: 3000,
```

## Author

👤 **Savio Resende**

* Website: https://savioresende.com.br
* GitHub: [@lotharthesavior](https://github.com/lotharthesavior)

## Show your support

Give a ⭐️ if this project helped you!

## 📝 License

Copyright © 2022 [Savio Resende](https://github.com/lotharthesavior).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
Updates should follow the [Keep a CHANGELOG](https://keepachangelog.com/) principles.

## [1.2.1] - 2023-03-21

### Fixed

- Fixed reconnection functionality. The previous package had some constraints to browsers.

### Added

- Callback for closing connections events.

## [1.2.0] - 2022-09-24

### Added

- Adjusted Client to the latest version of Socket Conveyor - changed the listen command.


index.js

class Conveyor {
    constructor(options) {
        this.options = {
            ...{
                protocol: 'ws',
                uri: '127.0.0.1',
                port: 8000,
                query: '',
                channel: null,
                listen: null,
                onOpen: this.onOpen.bind(this),
                onReady: () => {},
                onMessage: () => {}, // Message handler for only the data portion.
                onRawMessage: () => {}, // Message handler for the whole incoming object.
                onClose: () => this.onClose.bind(this),
                onCloseCallback: () => {},
                onError: () => {},
                reconnect: false,
                reconnectDelay: 5000,
                healthCheckInterval: 3000,
            },
            ...options
        };

        this.start();

        if (this.options.reconnect) {
            this.healthCheckInterval = setInterval(
                () => this.isClosed() ? this.onClose() : () => {},
                this.options.healthCheckInterval
            );
        }
    }

    isClosed() {
        return this.ws === null
            || 2 === this.ws.readyState // closing
            || 3 === this.ws.readyState // close
    }

    start() {
        this.ws = new WebSocket(this.options.protocol + '://' + this.options.uri + ':' + this.options.port + this.options.query);
        this.bindEvents();
    }

    bindEvents() {
        if (this.isClosed()) return;
        this.ws.onopen = this.options.onOpen;
        this.ws.onclose = this.options.onClose;
        this.ws.onerror = this.options.onError;
        this.ws.onmessage = this.baseOnMessage.bind(this);
    }

    onOpen(e) {
        this.connectChannel();
        this.addListeners();
        this.options.onReady();
    }

    onClose(e) {
        this.options.onCloseCallback();
        if (!this.options.reconnect) return;
        this.ws = null;
        setTimeout(() => this.start(), this.options.reconnectDelay);
    }

    baseOnMessage(e) {
        this.options.onRawMessage(e.data);
        const parsedData = JSON.parse(e.data);
        this.options.onMessage(parsedData.data);
    }

    send(message, action) {
        if (typeof action === 'undefined') {
            action = 'base-action';
        }

        this.rawSend(JSON.stringify({
            'action': action,
            'data': message,
        }));
    }

    rawSend(message) {
        if (this.isClosed()) return;
        this.ws.send(message);
    }

    connectChannel() {
        if (this.options.channel === null) {
            return;
        }

        this.rawSend(JSON.stringify({
            'action': 'channel-connect',
            'channel': this.options.channel,
        }));
    }

    addListeners() {
        if (this.options.listen === null) {
            return;
        }

        if (this.options.listen.constructor !== Array) {
            console.error('"listen" option must be an array.');
            return;
        }

        this.options.listen.forEach((action) => this.listen(action));
    }

    assocUser(userId) {
        this.rawSend(JSON.stringify({
            'action': 'assoc-user-to-fd-action',
            'userId': userId,
        }));
    }

    listen(action) {
        this.rawSend(JSON.stringify({
            'action': 'add-listener',
            'listen': action,
        }));
    }
}

export default Conveyor;
